Examples/GraphColoring.lean

```diff
@@ -0,0 +1,232 @@
+  [propfun|
+    (¬ ({blue    u} ∧ {blue   v})) ∧
+    (¬ ({red     u} ∧ {red    v})) ∧
+    (¬ ({green   u} ∧ {green  v})) ∧
+    (¬ ({yellow  u} ∧ {yellow v}))
+  ]
+
+def adjacentVertexesGetDifferentColors' (u v : Fin n) : PropFun (ColorVars n) :=
+  ((blue   u)ᶜ ⊔ (blue   v)ᶜ) ⊓
+  ((red    u)ᶜ ⊔ (red    v)ᶜ) ⊓
+  ((green  u)ᶜ ⊔ (green  v)ᶜ) ⊓
+  ((yellow u)ᶜ ⊔ (yellow v)ᶜ)
+
+-- Each edge has different colors
+def edgeConstraints (G : Graph n) : PropPred (ColorVars n) := fun τ =>
+  ∀ (u v), G.Adj u v → τ ⊨ adjacentVertexesGetDifferentColors u v
+
+-- The graph coloring problem
+def graphColoring (G : Graph n) : PropPred (ColorVars n) := fun τ =>
+  (τ |> vertexColorConstraints n) ∧
+  (τ |> edgeConstraints G)
+
+------------------------------------------------------------------------
+-- Now we express the graph coloring problem into a CNF
+/-! # CNF -/
+
+open Encode VEncCNF LitVar
+
+abbrev L (n : Nat) := Literal (ColorVars n)
+abbrev VCnf (n : Nat) := VEncCNF (L n) Unit
+
+def vertexColorClause (v : Fin n) : VCnf n (eachVertexGetsAColor v) :=
+  (addClause #[mkPos <| blue v, mkPos <| red v, mkPos <| green v, mkPos <| yellow v])
+  |> mapProp (by
+    simp [eachVertexGetsAColor, Clause.toPropFun]
+  )
+
+def vertexColorClauses (n : Nat) : VCnf n (vertexColorConstraints n) :=
+  ( let U := (Array.finRange n)
+    for_all U fun v =>
+      vertexColorClause v)
+  |> mapProp (by ext τ; apply forall_congr'; simp)
+
+def adjacentVertexesClauses (G : Graph n) [DecidableRel G.Adj] : VCnf n (edgeConstraints G) :=
+  ( let U := (Array.finRange n)
+    for_all U fun u =>
+    for_all U fun v =>
+    VEncCNF.guard (u ≠ v) fun _ =>
+    VEncCNF.guard (G.Adj u v) fun _ =>
+      seq[
+        addClause #[mkNeg <| blue   u, mkNeg <| blue   v],
+        addClause #[mkNeg <| red    u, mkNeg <| red    v],
+        addClause #[mkNeg <| green  u, mkNeg <| green  v],
+        addClause #[mkNeg <| yellow u, mkNeg <| yellow v]
+      ])
+  |> mapProp (by
+    ext τ
+    simp [edgeConstraints, adjacentVertexesGetDifferentColors, Clause.toPropFun]
+    apply forall_congr'; intro u
+    apply forall_congr'; intro v
+    aesop)
+
+def graphColoringCNF (G : Graph n) [DecidableRel G.Adj] : VCnf n (graphColoring G) :=
+  (seq[
+    vertexColorClauses n,
+    adjacentVertexesClauses G
+  ]).mapProp (by simp [graphColoring]; rfl)
+
+------------------------------------------------------------------------
+
+-- Now we prove that if there is a valid coloring, then we can produce an
+-- assignment on the variables for the graph
+
+def coloringAssignment (C : Coloring n 4) : PropAssignment (ColorVars n) :=
+  fun
+  | .blue   v => C v = ⟨0, by decide⟩
+  | .red    v => C v = ⟨1, by decide⟩
+  | .green  v => C v = ⟨2, by decide⟩
+  | .yellow v => C v = ⟨3, by decide⟩
+
+example (f : ℕ → Prop) (p : Fin 3) (h0 : f 0) (h1 : f 1) (h2 : f 2) : f p.val := by
+  fin_cases p; simp
+  all_goals assumption
+
+theorem splitFin4 (n : Fin 4) :
```

Comment:
`fin_cases n` should work directly in context

Reply:
You would expect it to. However, calling `fin_cases` on the output of the coloring function encounters a Lean error `unexpected syntax: fin_cases (C v)`. Putting a `let Cv := C v` beforehand gives a slightly different error:
```
tactic 'cases' failed, nested error:
dependent elimination failed, failed to solve equation
  { val := 0, isLt := (_ : 0 < 4) } = Cv
```
In my attempts to get a proof ready for formal cookies, externalizing the `fin_cases` in this theorem was the only thing that worked. I welcome additional changes in the future.